### ISSUE TITLE
Asciidoc Techdocs support

### ISSUE THREAD

**Issue by Hanyman8** [backstage/backstage#13402]
Hello,

we have our complete domunentation of microservices in adoc format. Would it be possible to add adoc format instead of just markdown using mkdocs.

One of the solution for us would be using our generated html from adoc (we are able to do it in our CI/CD)

## Feature Suggestion

Asciidoc support

## Context

We would like to serve our html generated from our asciidocs

**Comment by Hanyman8** [comment]
Hello, is there a way to serve our own html in the techdocs? Thanks

**Comment by tomb50** [comment]
We are also currently looking into adopting Backstage and have existing adoc documentation. We are trying to compare Backstage + referncing Antora vs what could be possible with pure Backstage Techdocs. 

Serving the HTML would indeed help us, even as an interim

**Comment by snowe2010** [comment]
@tomb50 we're in the exact same position. Antora, but others want to move to backstage. Backstage is great, but being dependent on markdown is quite terrible. There are several PRs that have been closed that would have accomplished this, maybe we can take them and add the support ourselves, since it seems like they're in no rush to add it. https://github.com/backstage/backstage/pull/9385

**Comment by Okeanos** [comment]
Yeah, so I am still interested in that.

**Comment by jarekkar** [comment]
Also interested in.

**Comment by Sroca3** [comment]
Same. Asciidoc is the default for spring-rest-docs and a majority of spring documentation is in asciidoc as far as I'm aware so it's a better option than markdown IMO.

**Comment by Okeanos** [comment]
Still relevant and waiting for a solution to this.

**Comment by ikatlinsky** [comment]
Still relevant, highly desired feature

**Comment by Okeanos** [comment]
Would still love to see this added.

**Comment by pwright** [comment]
I also had this requirement, but worked around it by generating html with no stylesheet [1] and an antora ui that  only produces body html, then renaming the html files as '.md', mkdocs was fine with the result

[1] https://docs.asciidoctor.org/asciidoctor/latest/html-backend/default-stylesheet/

**Comment by remast** [comment]
We' d also very much appreciate this.👍

**Comment by ikatlinsky** [comment]
Still highly desired

**Comment by JanGroot** [comment]
Yup, I want it as well

**Comment by StephaneRob** [comment]
As a POC, I created a s3 proxy plugin to server static documentation (storybook, ex_doc, ....) directly from backstage and avoid modifying the build to feat with TechDocs system.

**Comment by snowe2010** [comment]
@pwright @StephaneRob, can you both detail your solutions in a bit more detail to allow others to try and get it working as well? I think it would be highly appreciated.

**Comment by snowe2010** [comment]
Ok, I tried to do what they mentioned above, this is the solution I came up with. 

```kotlin
import org.asciidoctor.gradle.jvm.AsciidoctorTask // had to command click this after syncing to jump to the source and then intellij was able to find it

plugins {
	// ...
    id("org.asciidoctor.jvm.convert") version "4.0.1"
}

tasks.named<AsciidoctorTask>("asciidoctor") {
    outputOptions {
        backends("html5", "docbook")
    }

    sourceDir("docs/modules/ROOT")
    resources(delegateClosureOf<CopySpec> {
        from("docs/modules/ROOT/pages/resources") {
            include("*.png")
            include("**/*.png")
        }

        into("./resources")
    })

    options(mutableMapOf("standalone" to false))
    copyResourcesOnlyIf("html5")
    useIntermediateWorkDir()
    
    asciidoctorj {
        modules {
            diagram.use()
        }
    }

    doLast {
        copy {
            from("build/docs/asciidoc/html5")
            into("docs/generated-html/")
            rename("""(.*).html""", "$1.md")
        }
    }
}
```

My original directory structure looked like this

![image](https://github.com/backstage/backstage/assets/1534535/3bed0621-6fff-4e6d-abf9-8001ed86574e)

where all my antora docs and backstage component yamls are together.

**Comment by pwright** [comment]
@snowe2010 glad to see it worked out. TLDR; for folks `rename("""(.*).html""", "$1.md")` allowed Tyler to insert rendered AsciiDoc into mkdocs.

**Comment by snowe2010** [comment]
@pwright that and the `standalone` option _must_ be set to `false`

**Comment by StephaneRob** [comment]
On my side, I wanted to serve static documentation without modifying them. For this I did the following:

1. Build a front end plugin to add an entity route. This view has a selector to choose the doc to load for a given component and an iframe to load it (backend plugin routes)
2. Build a back end plugin with a catch all route to load html, css, js, font...proxying request to s3 bucket. (with a redis cache)

I needed:
- to have a convention of where i store my static doc
- make sure that assets are loaded with the correct url to proxy request to s3 using the same catch all route than the initial html

More work but can embed any type of static doc inside backstage component.

**Comment by AlexSneedMiller** [comment]
@StephaneRob are you still able to leverage the search functionality within Backstage in your s3 POC?

**Comment by snowe2010** [comment]
Not stale, just waiting...

**Comment by Hanyman8** [comment]
Thanks to everyone to keep this issue open. We would also still use this feature in our project.

**Comment by pwright** [comment]
@wanisfahmyDE that looks cool.. Given that you cannot accept contributions in the intermediate markdown format, are there advantages to converting to markdown rather than html as a workaround?

**Comment by wanisfahmyDE** [comment]
hey @pwright, not sure if I got your question right but the `techdocs-cli` generates htmls from the markdown eventually and this is what is published to S3 when running `techdocs-cli publish`.

Initially we tried achieving the same with pandoc and asciidoctor but this caused several issues like titles not appearing, incomplete table of contents and also broken PlantUMLs. Not to mention, maintaining a long bash script to do all these conversions.
downdoc and asciidoc-reducer kind of made it a "one liner(ish)" solution

**Comment by mojavelinux** [comment]
> Admontions with the paragraph syntax display a backslash \ at the end of the headline

This is hard line break. If you prefer not to use that style, pass `-a markdown-line-break` to the downdoc command.

See https://github.com/opendevise/downdoc/blob/main/test/downdoc-test.js#L1107-L1121

If you have other question about how downdoc works, you're welcome to ask in the project chat at https://chat.asciidoctor.org.

Naturally, not all features of AsciiDoc will be supported since Markdown is a simpler language than AsciiDoc.

**Comment by snowe2010** [comment]
@pwright yeah the html solution you initially suggested and I implemented for us in gradle has been working great.

**Comment by snowe2010** [comment]
not stale...

**Comment by thrix** [comment]
Maybe we could at least document these solutions for the teams having the majority of their docs in Asciidoc? I can contribute if anybody thinks this would be useful.

**Comment by thrix** [comment]
Would also like to hear if anybody had some luck converting Antora project to techdocs.

**Comment by he-mi** [comment]
A documented (and supported) solution would be great. Especially in combination with Antora.

**Comment by tomb50** [comment]
Not stale

**Comment by fredrik-forssen-cint** [comment]
Asciidoc is much nicer for documentation than Markdown, having this supported by backstage in some regard (preferably integrated with something like Antora) would be amazing.

**Comment by Okeanos** [comment]
Still a relevant feature request that I (and I guess many others) would like to see implemented.

**Comment by papanito** [comment]
Still relevant

**Comment by ganievs** [comment]
up

**Comment by bwinistoerfer** [comment]
Our company is using asciidoc as basis for its engineering handbook. This prevents us to move to the Backstage Techdocs which is really a pain. As asciidoc is widely used Backstage should really support it to allow more content in the Techdocs.

**Comment by Hanyman8** [comment]
Hello, any progress?

**Comment by freben** [comment]
No, this is not being worked on by the maintainers. It's now labeled as contrib-needed for this reason.

**Comment by pwright** [comment]
contrib to mkdocs? or here?

maybe folks on this thread would be interested in the state of mkdocs (underlying techdocs):
https://github.com/squidfunk/mkdocs-material/discussions/8461

**Comment by camilaibs** [comment]
Hi @pwright, it would be great to start by **proposing an RFC** that outlines how to make the TechDocs plugin support more documentation formats, such as AsciiDoc. A good starting point would likely involve:

1.  Investigating the best way to integrate `asciidoctor` (or a similar tool) into the TechDocs generator step.
2.  Potentially exploring a custom TechDocs generator implementation if tight integration with the MkDocs build process proves too complex.

In your investigation, you could also explore if a **plugin for MkDocs** that converts AsciiDoc to Markdown would be a viable solution. We've seen similar patterns in use; for example, some Backstage adopters created a converter that transforms Protobuf definition files into Markdown during build pipelines so their APIs are available in TechDocs. You might find other alternatives while writing the RFC. Your contribution in proposing a solution is more than welcome!

**Comment by Okeanos** [comment]
Bad Stale-Bot, this issue is still of interest to me.

**Comment by migueltercero** [comment]
Still relevant

**Comment by philhartung** [comment]
Hey everyone, just a heads-up for anyone still fighting with `downdoc` or HTML-renaming hacks: you might want to look into the [mkdocs-asciidoctor-backend](https://github.com/aireilly/mkdocs-asciidoctor-backend) plugin. 

It basically solves this without needing a messy workaround. Since TechDocs is just MkDocs under the hood, you can drop this plugin in and directly reference your `.adoc` files in your `mkdocs.yml`. It runs Asciidoctor to render the HTML, so you don't lose any native features like includes, tables, macros or even mermaid diagrams using the Asciidoctor Kroki plugin.

The only minor catch is styling. Asciidoctor generates its own HTML classes, so you'll need to inject some custom CSS to make those elements match the Backstage/Material theme. 

But once the CSS is set up, it works really well and feels like a native integration.

**Comment by fredrik-forssen-cint** [comment]
@philhartung Do you have any examples available on how to get that working?

**Comment by philhartung** [comment]
@fredrik-forssen-cint Sure, here is a minimal example to get you started:

**Prerequisites**

*(Note: You will need to add these to your CI/CD pipeline or your custom TechDocs Docker image, depending on how you generate your docs.)*
- Asciidoctor: `apt-get install asciidoctor` (or see [Asciidoctor Installation](https://docs.asciidoctor.org/asciidoctor/latest/install/))
- Asciidoctor Plugins (optional): `gem install asciidoctor-kroki rouge`
- MkDocs Asciidoctor Backend: `pip install mkdocs-asciidoctor-backend`

**mkdocs.yml**
```yaml
site_name: 'Asciidoc Example'

nav:
  - Home: index.adoc

plugins:
  - techdocs-core
  - asciidoctor_backend:
      attributes:
        backend: html5
        # Config for source highlighting with rouge
        source-highlighter: rouge
        rouge-css: 'class@'

# Custom Styling: The stylesheet name MUST follow the pattern main.[hash].min.css 
# or it will be stripped out by the Backstage sanitizer!
# For production use, it is highly recommended to add this custom styling directly to your Backstage app theme instead.
extra_css:
  - stylesheets/main.12345678.min.css
```

My custom stylesheet can be found here: [Gist](https://gist.github.com/philhartung/3a115e1b13d70e7b2ad176432952786d)

**Comment by Okeanos** [comment]
Bad Stale-Bot, this issue is still of interest to me.

**Comment by awanlin** [comment]
Hi! There's clearly a lot of desire for this but we aren't looking to support formats outside of Markdown in TechDocs.

We'd be open to a custom plugin to support this that could be contributed in the Community Plugins repo: https://github.com/backstage/community-plugins

We'd also be open to making some of the suggested workaround easier to do or getting them documented as a tutorial.

But the general direction here is that we very much are looking to the community to contribute this, again 92 👍 indicates that there has to be a few people willing to do so. I'm willing to help support and move forward any contributions for this that do come in.

**Comment by fredrik-forssen-cint** [comment]
Bad robot, we still want asciidocs in our techdocs.

**Comment by awanlin** [comment]
Hi @fredrik-forssen-cint and @bwinistoerfer and @trumme21, maybe you missed my comment: https://github.com/backstage/backstage/issues/13402#issuecomment-4927539987. Are you willing to help? If not then sadly the `stale` tag is valid, the community simply isn't able to support this.
